Game/RainMeadow.EntityHooks.cs

```diff
@@ -165,6 +166,32 @@ private void RoomRealizer_RealizeAndTrackRoom(On.RoomRealizer.orig_RealizeAndTra
             orig(self, room, actuallyEntering);
         }
 
+        // Prevent creatures from entering a room if their online counterpart has not yet entered!
+        private bool ShortcutHandlerOnVesselAllowedInRoom(On.ShortcutHandler.orig_VesselAllowedInRoom orig, ShortcutHandler self, ShortcutHandler.Vessel vessel)
+        {
+            var result = orig(self, vessel);
+            if (self.game.session is not OnlineGameSession) return result;
+
+            var absCrit = vessel.creature.abstractCreature;
+            OnlineEntity.map.TryGetValue(absCrit, out var onlineEntity);
+            if (onlineEntity.owner.isMe) return result; // If entity is ours, game handles it normally.
+            
+            if (onlineEntity.entity.Room != vessel.room) result = false; // If OnlineEntity is not yet in the room, keep waiting.
```

Comment:
This is unnecessary because AllConnectedObjects includes self. Could be left in as a fast-fail though.

However the condition it is checking for is wrong. It should be checking for onlineEntity.Room.absRoom != vessel.room